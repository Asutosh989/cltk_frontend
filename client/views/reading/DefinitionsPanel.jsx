
DefinitionsPanel = React.createClass({

  propTypes: {
    toggleDefinitions: React.PropTypes.bool,
    textNodes: React.PropTypes.array
  },
  getDefaultProps() {
    return {
      toggleDefinitions: false,
      textNodes: []
    };
  },

  // This mixin makes the getMeteorData method work
  mixins: [ReactMeteorData],

  getMeteorData() {
    words = [];
    this.props.textNodes.map((textNode, i) => {
      definitions = {};
      Wordforms.find({texts: textNode._id}).fetch().map((wordform) => {
        if(definitions[wordform.word] == null){
          definitions[wordform.word] = [];
        }
        definition = Definitions.findOne({_id: wordform.definitions});
        if(definition != undefined){
          definitions[wordform.word].push(definition);
        }
      });
      for(key in definitions) {
        word = {};
        word["index"] = i;
        word["lemma"] = key;
        word["definitions"] = definitions[key];
        words.push(word);
      }
      });

    return {
      words : words
    };

  },


  renderDefinitions(){
    /*var words = [
      {
        _id : 123,
        lemma: "arma",
        definitions: [
          {
            orthography: ["armum", "armi"],
            senses: ["arms (pl.)", "weapons", "armor", "shield", "close fighting weapons", "equipment", "force"],
            inflections: [
              {
                pos: "noun",
                ending: "a",
                declension: {
                  title: "2nd",
                  "case": "nominative"
                },
                gender: "neuter",
                number: "plural"
              }, {
                pos: "noun",
                ending: "a",
                declension: {
                  title: "2nd",
                  "case": "vocative"
                },
                gender: "neuter",
                number: "plural"
              }, {
                pos: "noun",
                ending: "a",
                declension: {
                  title: "2nd",
                  "case": "accusative"
                },
                gender: "neuter",
                number: "plural"
              }
            ]
          }, {
            orthography: ["armo", "armare", "aramvi", "armatus"],
            senses: ["equip", "fit with armor", "arm", "strengthen", "rouse", "stir", "incite war", "rig (ship)"],
            inflections: [
              {
                pos: "verb",
                ending: "a",
                conjugation: "1st",
                person: "2nd",
                number: "singular",
                tense: "present",
                voice: "active",
                mood: "imperative"
              }
            ]
          }
        ]
      }, {
        _id : 124,
        lemma: "virum",
        definitions: [
          {
            orthography: ["virum", "viri"],
            senses: ["man", "husband", "hero", "person of courage", "honor", "nobility"],
            inflections: [
              {
                pos: "noun",
                ending: "um",
                declension: {
                  title: "2nd",
                  "case": "accusative"
                },
                gender: "masculine",
                number: "singular"
              }
            ]
          }, {
            orthography: ["vis", "viris"],
            senses: ["strength (bodily) (pl.)", "force", "power", "might", "violence", "resources", "large body"],
            inflections: [
              {
                pos: "noun",
                ending: "um",
                declension: {
                  title: "3rd",
                  "case": "genitive"
                },
                gender: "feminine",
                number: "plural"
              }
            ]
          }
        ]
      }
    ];
    */
    return this.data.words.map((word, i) => {
      return <DefinitionWord
        key={i}
        word={word} />;
    });

  },

  render() {
     return (
        <div className={(this.props.toggleDefinitions)? "slide-visible modal-panel definitions-panel paper-shadow"
          :"modal-panel definitions-panel paper-shadow"}>
          <div className="modal-panel-inner definitions-panel-inner">
            <div className="definitions panel-items">
              {this.renderDefinitions()}
            </div>

            {this.data.words.length === 0 ?
                <span className="no-results no-results-definitions">No definitions available.</span>
              : null }
          </div>
         </div>


     );
   }

});
