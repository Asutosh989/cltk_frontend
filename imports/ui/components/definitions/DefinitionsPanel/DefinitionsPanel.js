import React from 'react';
import ReactList from 'react-list';
import PropTypes from 'prop-types';
import autoBind from 'react-autobind';
import { createContainer } from 'meteor/react-meteor-data';
import baseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import TextField from 'material-ui/TextField';

import Definitions from '/imports/api/collections/definitions';
import Wordforms from '/imports/api/collections/wordforms';

class DefinitionsPanel extends React.Component {

	constructor(props) {
		super(props);

		this.state = {
			searchText: '',
		};

		autoBind(this);
	}

	getDefaultProps() {
		return {
			toggleDefinitions: false,
			textNodes: [],
		};
	}


	getChildContext() {
		return { muiTheme: getMuiTheme(baseTheme) };
	}

	handleChange(event) {
		this.setState({ searchText: event.target.value.toLowerCase() });
	}

	renderDefinitions() {
		return this.props.words.map((word, i) => (
			<DefinitionWord
				key={i}
				word={word}
			/>
		));
	}

	renderDefinition(index, key) {
		return (
			<DefinitionWord
				key={key}
				word={this.props.words[index]}
			/>
		);
	}

	render() {
		return (
			<div
				className={(this.props.toggleDefinitions) ?
				'slide-visible modal-panel definitions-panel paper-shadow' :
				'modal-panel definitions-panel paper-shadow'}
			>
				<div className="modal-panel-inner definitions-panel-inner">
					<TextField
						className="search-box" hintText="Search text"
						fullWidth
						onChange={this.handleChange}
					/>
					<div className="definitions panel-items" >
						<ReactList
							itemRenderer={this.renderDefinition}
							length={this.props.words.length}
							type="variable"
						/>

					</div>

					{this.props.words.length === 0 ?
						<span className="no-results no-results-definitions">No definitions available.</span>
						:
						null
					}
				</div>
			</div>
		);
	}
}

DefinitionsPanel.propTypes = {
	toggleDefinitions: PropTypes.bool,
	textNodes: PropTypes.array,
};

DefinitionsPanel.childContextTypes = {
	muiTheme: PropTypes.object.isRequired,
};


const DefinitionsPanelContainer = createContainer((props) => {
	const words = [];
	const textIds = [];
	const definitionIds = [];
	const definitions = {};
	let wordForms = [];
	props.textNodes.map((textNode) => {
		textIds.push(textNode._id);
		return true;
	});

	const handleWordforms = Meteor.subscribe('wordForms', textIds);
	if (handleWordforms.ready()) {
		wordForms = Wordforms.find({ word: { $regex: props.searchText } }).fetch();
		wordForms.map((wordForm) => {
			definitionIds.push(wordForm.definitions);
			return true;
		});
		const handleDefinitions = Meteor.subscribe('definitions', definitionIds);
		if (handleDefinitions.ready()) {
			wordForms.map((wordForm) => {
				if (definitions[wordForm.word] == null) {
					definitions[wordForm.word] = [];
				}
				definition = Definitions.findOne({ _id: wordForm.definitions });
				if (definition !== undefined) {
					definitions[wordForm.word].push(definition);
				}
				return true;
			});
			Object.keys(definitions).forEach(key => {
				word = {};
				word.lemma = key;
				word.definitions = definitions[key];
				words.push(word);
			});
		}
	}
	return {
		words,
	};
}, DefinitionsPanel);

export default DefinitionsPanelContainer;
