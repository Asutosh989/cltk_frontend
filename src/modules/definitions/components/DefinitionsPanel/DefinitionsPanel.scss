
.definition {

  h4.word, .definition-link {
    display:inline-block;
    vertical-align:top;
  }

  h4.word {
    margin:0;
  }

  .definition-link {
    margin:7px 10px;

  }

  .meaning {
    width:98%;
    margin:0 auto;
    padding:5px 0 10px;
  }

  .root, .meaning-definition, .form {
    display:block;
    color:@text-color;

  }

  .root {
    font-size:16px;
    font-weight:bold;
    padding:0 0 3px;

  }

  .meaning-definition {
    font-size:16px;
    line-height:1.5em;
    padding:0 0 3px;

  }

  .form {
    font-size:14px;
    margin-left:10px;
  }



}
