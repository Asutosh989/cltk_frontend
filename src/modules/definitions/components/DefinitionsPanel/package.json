{
  "name": "DefinitionsPanel",
  "version": "0.0.1",
  "private": true,
  "main": "./DefinitionsPanel.js"
}
