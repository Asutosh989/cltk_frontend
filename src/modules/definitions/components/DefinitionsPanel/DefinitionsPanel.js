import React from 'react';
import ReactList from 'react-list';
import PropTypes from 'prop-types';
import autoBind from 'react-autobind';
import baseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import TextField from 'material-ui/TextField';


import './DefinitionsPanel.css';


class DefinitionsPanel extends React.Component {
	static defaultProps = {
		toggleDefinitions: false,
		textNodes: [],
	}

	constructor(props) {
		super(props);

		this.state = {
			searchText: '',
		};

		autoBind(this);
	}

	getChildContext() {
		return { muiTheme: getMuiTheme(baseTheme) };
	}

	handleChange(event) {
		this.setState({ searchText: event.target.value.toLowerCase() });
	}

	renderDefinitions() {
		return this.props.words.map((word, i) => (
			<DefinitionWord
				key={i}
				word={word}
			/>
		));
	}

	renderDefinition(index, key) {
		return (
			<DefinitionWord
				key={key}
				word={this.props.words[index]}
			/>
		);
	}

	render() {
		return (
			<div
				className={(this.props.toggleDefinitions) ?
				'slide-visible modal-panel definitions-panel paper-shadow' :
				'modal-panel definitions-panel paper-shadow'}
			>
				<div className="modal-panel-inner definitions-panel-inner">
					<TextField
						className="search-box" hintText="Search text"
						fullWidth
						onChange={this.handleChange}
					/>
					<div className="definitions panel-items" >
						<ReactList
							itemRenderer={this.renderDefinition}
							length={this.props.words.length}
							type="variable"
						/>

					</div>

					{this.props.words.length === 0 ?
						<span className="no-results no-results-definitions">No definitions available.</span>
						:
						null
					}
				</div>
			</div>
		);
	}
}

DefinitionsPanel.propTypes = {
	toggleDefinitions: PropTypes.bool,
	textNodes: PropTypes.array,
};

DefinitionsPanel.childContextTypes = {
	muiTheme: PropTypes.object.isRequired,
};


export default DefinitionsPanel;
