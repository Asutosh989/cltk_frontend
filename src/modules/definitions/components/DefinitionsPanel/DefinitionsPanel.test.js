import React from 'react';
import { shallow } from 'enzyme';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';

// component:
import DefinitionsPanel from './DefinitionsPanel';

jest.mock('material-ui/TextField/TextField');
jest.mock('react-list');

describe('DefinitionsPanel', () => {
	it('renders correctly', () => {

		const wrapper = shallow(
			<MuiThemeProvider>
				<DefinitionsPanel
					textNodes={[]}
					words={[]}
					/>
			</MuiThemeProvider>
			);
		expect(wrapper).toBeDefined();
	});
});
