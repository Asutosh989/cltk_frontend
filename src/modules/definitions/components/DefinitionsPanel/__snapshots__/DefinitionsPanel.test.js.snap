// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefinitionsPanel renders correctly 1`] = `
<div
  className="modal-panel definitions-panel paper-shadow"
>
  <div
    className="modal-panel-inner definitions-panel-inner"
  >
    <div
      className="definitions panel-items"
    />
    <span
      className="no-results no-results-definitions"
    >
      No definitions available.
    </span>
  </div>
</div>
`;
